votingpool/input_selection.go

```diff
@@ -0,0 +1,286 @@
+/*
+ * Copyright (c) 2015 Conformal Systems LLC <info@conformal.com>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+package votingpool
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcd/wire"
+	"github.com/btcsuite/btcutil"
+	"github.com/btcsuite/btcwallet/txstore"
+)
+
+const eligibleInputMinConfirmations = 100
+
+// Credit is an abstraction over txstore.Credit used in the construction of
+// voting pool withdrawal transactions.
+type Credit interface {
+	fmt.Stringer
+	TxSha() *wire.ShaHash
+	OutputIndex() uint32
+	Address() WithdrawalAddress
+	Amount() btcutil.Amount
+	OutPoint() *wire.OutPoint
+	TxOut() *wire.TxOut
+}
+
+// credit implements the Credit interface.
+type credit struct {
+	txstore.Credit
+	addr WithdrawalAddress
+}
+
+// newCredit initialises a new credit.
+func newCredit(c txstore.Credit, addr WithdrawalAddress) *credit {
+	return &credit{Credit: c, addr: addr}
+}
+
+func (c *credit) String() string {
+	return fmt.Sprintf("credit of %v to %v", c.Amount(), c.Address())
+}
+
+// TxSha returns the sha hash of the underlying transaction.
+func (c *credit) TxSha() *wire.ShaHash {
+	return c.Credit.TxRecord.Tx().Sha()
+}
+
+// OutputIndex returns the outputindex of the ouput in the underlying
+// transaction.
+func (c *credit) OutputIndex() uint32 {
+	return c.Credit.OutputIndex
+}
+
+// Address returns the voting pool address.
+func (c *credit) Address() WithdrawalAddress {
+	return c.addr
+}
+
+// Compile time check that credit implements Credit interface.
+var _ Credit = (*credit)(nil)
+
+// byAddress defines the methods needed to satisify sort.Interface to sort a
+// slice of Credits by their address.
+type byAddress []Credit
+
+func (c byAddress) Len() int      { return len(c) }
+func (c byAddress) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
+
+// Less returns true if the element at positions i is smaller than the
+// element at position j. The 'smaller-than' relation is defined to be
+// the lexicographic ordering defined on the tuple (SeriesID, Index,
+// Branch, TxSha, OutputIndex).
+func (c byAddress) Less(i, j int) bool {
+	if c[i].Address().seriesID < c[j].Address().seriesID {
+		return true
+	}
+
+	if c[i].Address().seriesID == c[j].Address().seriesID &&
```

Comment:
Super minor thing, but I wonder if rather than rechecking each of these conditions multiple times, if it would be better to do something like:

```Go
if c[i].Address().seriesID < c[j].Address().seriesID {
	return true
}
if c[i].Address().seriesID > c[j].Address().seriesID {
	return false
}

// The seriesID are equal, so compare index.
if c[i].Address().index < c[j].Address().index {
	return true
}
if c[i].Address().index > c[j].Address().index {
	return false
}

// The seriesID and index are equal, so compare branch.
if c[i].Address().branch < c[j].Address().branch {
	return true
}
if c[i].Address().branch > c[j].Address().branch {
	return false
}

// The seriesID, index, and branch are equal, so compare hash.
txidComparison := bytes.Compare(c[i].TxSha().Bytes(), c[j].TxSha().Bytes())
if txidComparison < 0 {
	return true
}
if txidComparison > 0 {
	return false
}

// The seriesID, index, branch, and hash are equal, so compare index.
if c[i].OutputIndex() < c[j].OutputIndex() {
	return true
}

return false
```